src/include/processor/operator/partitioner.h

```diff
@@ -102,21 +112,21 @@ class Partitioner : public Sink {
 
     std::unique_ptr<PhysicalOperator> clone() final;
 
-    static void initializePartitioningStates(std::vector<std::unique_ptr<PartitioningInfo>>& infos,
+    static void initializePartitioningStates(std::unique_ptr<PartitionerDataInfo>& dataInfo,
         std::vector<std::unique_ptr<PartitioningBuffer>>& partitioningBuffers,
         std::vector<common::partition_idx_t> numPartitions);
 
 private:
-    common::DataChunk constructDataChunk(const std::vector<DataPos>& columnPositions,
-        const std::vector<common::LogicalType>& columnTypes, const ResultSet& resultSet,
-        const std::shared_ptr<common::DataChunkState>& state);
+    void evaluateData(const common::sel_t& numTuples);
+    common::DataChunk constructDataChunk(const std::shared_ptr<common::DataChunkState>& state);
     // TODO: For now, RelBatchInsert will guarantee all data are inside one data chunk. Should be
     //  generalized to resultSet later if needed.
     void copyDataToPartitions(common::partition_idx_t partitioningIdx,
         common::DataChunk chunkToCopyFrom);
 
 private:
     std::vector<std::unique_ptr<PartitioningInfo>> infos;
+    std::unique_ptr<PartitionerDataInfo> dataInfo;
```

Comment:
Not related to this PR but I guess you could also change unique_ptr of `PartitioningInfo` to object?